### PR TITLE
Embed long documents with openai

### PR DESCRIPTION
**Pull request by noamzbr**
Embed long documents using the recipe from https://github.com/openai/openai-cookbook/blob/main/examples/Embedding_long_inputs.ipynb

### CODE REVIEW

File: deepchecks/nlp/utils/text_embeddings.py [hunk 1]
```diff
@@ -9,14 +9,52 @@
...
+def iterate_batched(tokenized_text, chunk_length):
+    """Chunk text into tokens of length chunk_length."""
+    chunks_iterator = batched(tokenized_text, chunk_length)
+    yield from chunks_iterator
+
 
 def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
-                                 file_path: Optional[str] = 'embeddings.npy') -> np.array:
+                                 file_path: Optional[str] = 'embeddings.npy',
+                                 device: Optional[str] = None) -> np.array:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_embeddings.py:57]: we discussed an argument to control what to do with texts longer then limit and that by default will be average with warning/exception

Comment (Nadav-Barak) [reply deepchecks/nlp/utils/text_embeddings.py:57]: the warnings / exception should state with samples were longer then limit + what is the limit (which is based on selected model i guess)

File: deepchecks/nlp/utils/text_embeddings.py [hunk 2]
```diff
@@ -30,6 +68,8 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
         by using the command openai.api_key = YOUR_API_KEY
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_embeddings.py:68]: Need to add info on that process we are doing + ref link to open ai cook book

File: deepchecks/nlp/utils/text_embeddings.py (hunk 1 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_embeddings.py:57]: Can't state which samples, can be too many. Will length of maximal one

File: deepchecks/nlp/utils/text_embeddings.py (hunk 2 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_embeddings.py:70]: Added

File: deepchecks/nlp/text_data.py [hunk 3]
```diff
@@ -313,11 +314,24 @@ def calculate_builtin_embeddings(self, model: str = 'miniLM', file_path: str = '
             'open_ai': using the ADA model in the open_ai library. Requires an API key.
         file_path : str, default: 'embeddings.npy'
             The path to save the embeddings to.
+        device : str, default: None
+            The device to use for calculating the embeddings. If None, the default device will be used.
+        long_doc_averaging : str, default 'warn'
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:319]:
```suggestion
        long_doc_averaging : str, default ''average+warn''
```

File: deepchecks/nlp/text_data.py [hunk 4]
```diff
@@ -313,11 +314,24 @@ def calculate_builtin_embeddings(self, model: str = 'miniLM', file_path: str = '
...
+            How to handle long documents (docments that are longer than the model context window).
+            Can be either 'average+warn', 'average', 'truncate' or 'raise'. Averaging is done as described in
+            https://github.com/openai/openai-cookbook/blob/main/examples/Embedding_long_inputs.ipynb
+
+            Currently, applies only to the 'open_ai' model, as the 'miniLM' model can handle long documents.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:324]: Add a todo, i guess for really long documents it cant

File: deepchecks/nlp/text_data.py (hunk 3 above)
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:319]: Add a option of return None for long samples

File: deepchecks/nlp/text_data.py [hunk 5]
```diff
@@ -313,11 +314,24 @@ def calculate_builtin_embeddings(self, model: str = 'miniLM', file_path: str = '
             'open_ai': using the ADA model in the open_ai library. Requires an API key.
         file_path : str, default: 'embeddings.npy'
             The path to save the embeddings to.
+        device : str, default: None
+            The device to use for calculating the embeddings. If None, the default device will be used.
+        long_doc_averaging : str, default 'warn'
+            How to handle long documents (docments that are longer than the model context window).
+            Can be either 'average+warn', 'average', 'truncate' or 'raise'. Averaging is done as described in
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:321]: alternatives to consider: 
1. always warn
2. default is raise and if average is selected dont warn

File: deepchecks/nlp/utils/text_embeddings.py [hunk 6]
```diff
@@ -30,6 +70,17 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
...
     file_path : Optional[str], default 'embeddings.csv'
         If given, the embeddings will be saved to the given file path.
+    device : str, default None
+        The device to use for the embeddings. If None, the default device will be used.
+    long_doc_averaging : str, default 'warn'
+        How to handle long documents (docments that are longer than the model context window).
+         Can be either 'average+warn', 'average', 'truncate' or 'raise'.
+         Currently, applies only to the 'open_ai' model, as the 'miniLM' model can handle long documents. Averaging is
+         done as described in https://github.com/openai/openai-cookbook/blob/main/examples/Embedding_long_inputs.ipynb
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_embeddings.py:79]: same here but lets discuss in above thread

File: requirements/nlp-requirements.txt [hunk 7]
```diff
@@ -1,4 +1,6 @@
 seqeval>=1.0.0
 nltk>=3.4.0,<=3.6.7
 textblob>=0.17.1
-umap-learn
\ No newline at end of file
+umap-learn
+transformers>=4.0.0
+tiktoken; python_version >= '3.8'
```
Comment (Nadav-Barak) [inline requirements/nlp-requirements.txt:6]: must?

File: deepchecks/nlp/text_data.py (hunk 4 above)
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:324]: it can, the problem is just with open_ai

File: deepchecks/nlp/text_data.py (hunk 5 above)
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:321]: 1. Against - has to be an option to average without warning
2. I think you want things to "work" out of the box.

File: requirements/nlp-requirements.txt (hunk 7 above)
Comment (noamzbr) [reply requirements/nlp-requirements.txt:6]: yes

File: deepchecks/nlp/utils/text_embeddings.py (hunk 6 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_embeddings.py:79]: Not sure what is same here

File: deepchecks/nlp/text_data.py (hunk 5 above)
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:321]: (but without saying doc, see above comment)

File: deepchecks/nlp/utils/text_embeddings.py [hunk 8]
```diff
@@ -30,6 +70,18 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
         by using the command openai.api_key = YOUR_API_KEY
     file_path : Optional[str], default 'embeddings.csv'
         If given, the embeddings will be saved to the given file path.
+    device : str, default None
+        The device to use for the embeddings. If None, the default device will be used.
+    long_doc_averaging : str, default 'warn'
+        How to handle long documents (docments that are longer than the model context window).
+         Can be either 'average+warn', 'average', 'truncate' or 'raise'.
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_embeddings.py:77]: If you're describing the options later, no need to say them here

File: deepchecks/nlp/utils/text_embeddings.py [hunk 9]
```diff
@@ -53,21 +105,74 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
...
+                                  f'found is {max_doc_length} tokens. The document will be split into chunks and the '
+                                  f'embeddings will be averaged. To avoid this warning, set '
+                                  f'long_doc_averaging="average" or long_doc_averaging="truncate".')
+                elif long_doc_averaging == 'raise':
+                    raise ValueError(f'At least one document is longer than {max_tokens} tokens, which is the maximum '
+                                     f'context window handled by {model}. Maximal document '
+                                     f'length found is {max_doc_length} tokens. To avoid this error, set '
+                                     f'long_doc_averaging="average" or long_doc_averaging="truncate".')
+
+            batch_size = 500
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_embeddings.py:148]: should be configurable parameter

File: deepchecks/nlp/utils/text_embeddings.py [hunk 10]
```diff
@@ -53,21 +105,74 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
...
+
+            result_embeddings = []
+            idx = 0
+            for tokens_in_sample in encoded_texts:
+                text_embeddings = []
+                text_lens = []
+                while idx < len(chunk_lens) and sum(text_lens) < len(tokens_in_sample):
+                    text_embeddings.append(chunk_embeddings[idx])
+                    text_lens.append(chunk_lens[idx])
+                    idx += 1
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_embeddings.py:163]: Why the while loop? Can be just replaced with 2 list comprehensions... no?

File: deepchecks/nlp/utils/text_embeddings.py [hunk 11]
```diff
@@ -53,21 +105,74 @@ def calculate_builtin_embeddings(text: np.array, model: str = 'miniLM',
...
+
+                if long_doc_averaging == 'nan' and len(text_embeddings) > 1:
+                    text_embedding = np.ones_like(text_embeddings[0]) * np.nan
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_embeddings.py:166]: If I get this correctly, you still calculate the text embeddings for all the chunks even if you truncate, you just don't use it in the end? I don't think this is good, this wastes a lot of resources + the user pays for each API call.

File: deepchecks/nlp/utils/text_properties.py [hunk 12]
```diff
@@ -39,6 +39,7 @@
 MODELS_STORAGE = pathlib.Path(__file__).absolute().parent / '.nlp-models'
 FASTTEXT_LANG_MODEL = 'https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin'
 DEFAULT_SENTENCE_SAMPLE_SIZE = 300
+MAX_CHARS = 720  # Bert accepts max of 512 tokens, and experimentally 720 chars seems to work while 1024 doesn't
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_properties.py:42]: I mean.
Technically, a user can have a string where each character is a token.

File: deepchecks/nlp/utils/text_properties.py [hunk 13]
```diff
@@ -344,6 +345,21 @@ def subjectivity(text: str) -> float:
 def _predict(text: str, classifier, kind: str) -> float:
     """Return prediction of huggingface Pipeline classifier."""
     try:
+        # TODO: make this way smarter, and not just a hack. Count tokens, for a start. Then not just sample sentences.
+        # If text is longer than classifier context window, sample it:
+        if len(text) > MAX_CHARS:
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_properties.py:350]: Why not do the tokenization hack you have for open ai?

File: deepchecks/nlp/utils/text_embeddings.py (hunk 10 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_embeddings.py:163]: because of the sum len condition

File: deepchecks/nlp/utils/text_properties.py (hunk 12 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_properties.py:42]: yeah this is more a patch then a complete solution

File: deepchecks/nlp/text_data.py (hunk 5 above)
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:321]: I think that though seperating is more "complete", it's simpler and easier to have just one parameter controlling everything about the behaviour for long docs. Also, the logic is that if you explicitly set a method, you don't need a warning for it ever.

Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:321]: I really think this is too complicated.
And I get the warning, but Keep It Simple. I think this is a complicated solution that's not user friendly. Never seen a parameter that does 2 things like that, and that they are completely unrelated.

File: deepchecks/nlp/utils/text_embeddings.py (hunk 10 above)
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_embeddings.py:163]: But the text_lens can't be over the number of tokens in sample, can it? you run on batches anyway...

File: deepchecks/nlp/utils/text_properties.py (hunk 13 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_properties.py:350]: I'd have to use the specific tokenizer of each model here, decided to avoid for now...

**Review by nirhutnik — APPROVED** [review]

File: deepchecks/nlp/utils/text_properties.py (hunk 12 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_properties.py:42]: will do

**Pull request merged**